### ISSUE TITLE
Tokenization is time and space inefficient

### ISSUE THREAD

**Issue by kliron** [fastai/fastai#1998]
Going through the code in [transform.py](https://github.com/fastai/fastai/blob/master/fastai/text/transform.py) I cannot help but notice several opportunities for optimization for parallel execution. In its current form it would take >4 days to tokenize a 12Gb corpus on a 16-core/32 thread CPU (if it wouldn't run out of memory first as 36G RAM weren't enough). Writing a custom implementation reduced both the time to a little more than 4 hours and memory use by 2-3x. The code is mission-specific and write-once dirt so I'm feeling reluctant to share, but I'd gladly share the gist of it below.

1. In its current implementation the tokenization process is parallelized using the very inefficient `concurrent.futures.ProcessPoolExecutor`'s `map` function which creates Future objects where there is no good reason to. These are good for fine-grained control like progress reporting, cancelling etc, but are fairly heavy. In this case we are actually only interested in the returned tokens. `multiprocessing.Pool`'s `map` should perform considerably better. See [this SO post](https://stackoverflow.com/questions/18671528/processpoolexecutor-from-concurrent-futures-way-slower-than-multiprocessing-pool
) for more details.

2. A number of new processes are created to tokenize each bach of text. This means a new batch of processes need to be forked, initialize a fresh instance of Spacy Tokenizer and receive a fairly large chunk of text via IPC  every few seconds. This seems very inefficient. For small enough batches more time will be spent in fork-IPC-joining than on the actual work being done. Alternatively, there should be a number of long-lived tokenizer worker processes initialized at the beginning with a workload to process and each should process a stream of text with the more efficient `Tokenizer.pipe` function from [Spacy](https://spacy.io/api/tokenizer#pipe). 

Also, for large batches it is by leaps and bounds more efficient to have each process read its own batch from disk than to have some producer process provide it by IPC. Python's performance for reading large objects through IPC is atrocious (I don't know if this is Python-specific). See [this SO post](https://stackoverflow.com/questions/47085458/why-is-multiprocessing-queue-get-so-slow) for more context.

3. The current implementation requires enormous amounts of RAM for relatively small corpora (something like ~1Gb requires >24G RAM). Serializing the tokens and word counts from each tokenizer worker to disk, merging them after tokenization and (if needed) truncate the vocabulary and replace deleted instances with UNK in the tokenized text files is vastly more memory-efficient and can scale to much larger texts. 

I don't really know if the goal of the code as it exists today is to make it easier to "bring your own tokenizer" or to just make it small and understandable, and it is definitely nice to have functions like `.from_csv` or `.from_files` that "magically" do everything in one go, for demonstration purposes, but for more serious datasets, maybe breaking the process to more manageable pieces would be a better approach? 

[EDIT: Some demonstration]

Here is what processor utilization looks like with the current implementation:

![impl1](https://user-images.githubusercontent.com/626811/56718408-352fea80-672e-11e9-8c6c-aa90f1692327.png)

Here is what it should look like (running the example script from [here](https://github.com/kliron/example-multiprocess-tokenization-with-Spacy/blob/master/tokenization.py))

![impl2](https://user-images.githubusercontent.com/626811/56718481-5bee2100-672e-11e9-87a4-5ef0ed9eb90a.png)

**Comment by sgugger** [comment]
Thanks for the detailed analysis. The tokenization is written like this because it is enough for basic needs: you can tokenize and numericalize something as large as wikipedia with it. It's limited if you want larger corpuses, I agree. Will remember the tips about `multiprocessing.Pool` and we are already aware of `pipe` and will soon plug it in (when the current live course is finished).

The RAM issue isn't something we will address just yet. It requires a specific and very different tokenizer and numericalizer to dump things to disk as the work progresses then reading through it once to build the vocab, then another time to numericalize. I may look into this feature sometimes next month if there is sufficient demand for it.

In the meantime, leaving this issue open until I treat points 1 and 2 next week!

**Comment by bfarzin** [comment]
I have a PR coming that dumps to text file and then generates vocab and tokenizes.  Lowers the RAM usage as we pass through.

**Comment by kliron** [comment]
Edited the initial post with an example script. My use case is reading from files but it shouldn't be that difficult to spread row index ranges of a pandas frame or a CSV file instead of file names across N workers for example. For extremely large files, memory mapping could be used to avoid loading the same file in memory N times.

I'd also like to add that I think tokenization and numericalization are orthogonal processes and should not be merged into a single step. It is not uncommon to want to test different tokenization strategies, or to want to have raw tokens to use in other libraries, merge files from different sources, recreate/merge/split vocabularies, etc. The most common scenario is for example to want to keep capitalization say, for NER tasks, or not, say, for classification tasks. It would be ridiculous to be forced to re-process an enormous corpus just to have a different capitalization strategy.

**Comment by sgugger** [comment]
Yes I don't want numericalization and tokenization to be merged either.

**Comment by bfarzin** [comment]
... I will re-work my PR then. I had merged them to get it faster.  But I understand these points and can work around it.

**Comment by bfarzin** [comment]
I have made a lot of progress on this and want to put out a PR for the two of you to look at even though it is not fully completed yet.  I am using the pipe in the tokenizer and spawning processes as indicated. It does run a lot faster.  Much of the code is lifted from @kilron 's example.  I write to a set of files that can then be re-used for Numericalization.   I am curious if you have an opinion about the best place to put those files, right now I use `./` but clearly could put them anywhere. 

One problem I have encountered is not being able to update the `progress_bar` since this is spun out between multiple processes.  What I want to do is track a single one of the processes, but when I try to use `fastprogress` with `Process` something goes wrong (and I get no updates in a notebook and things don't seem to complete)  

Any ideas/suggestions on how to handle this case?

**Comment by sgugger** [comment]
You should look at `parallel` (or even use it directly) for updating the progressbar with multiprocess.

**Comment by kliron** [comment]
@bfarzin instead of a `Queue` you could use `multiprocess.Pool.imap` to start the workers and wrap the call in `tqdm()`. I'll update the example as soon as I have the time.

**Comment by bfarzin** [comment]
Thanks. I thought a second queue could work but this is all new to me. The example helps a lot.

**Comment by kliron** [comment]
Great. Just make sure to drain the reporting queue before starting to process the data queue as `Queue.get` is a blocking call.

**Comment by bfarzin** [comment]
Thanks for the detail.  I will see if I can get the latter one working as well.

**Comment by kliron** [comment]
To do that you would initialize the progress bar from within the main function and pass it as an argument to the worker processes (this argument would have a default value of `None`). Either initialize it with the number of objects the reporter process will have, in which case the reported number will be falsely low, or, initialize it with the total number and call `.update(n)` where `n` is the number of workers, in which case the reported progress is inaccurate but hopefully close to the truth.
In the latter case, you need to also pass the number n as yet another argument...

As you see, this quicky becomes a steaming pile of sh*t just to report some freaking progress. Observe how all of the above would be solved by creating a closure that would capture the progress bar object and the number of workers which you could pass to a selected process to call. But this is impossible because Callables are not picklable and therefore cannot be passed as Process arguments.

**Comment by bfarzin** [comment]
Good insight about the closure.  That makes sense.   I do agree it is work and slowness.  but I also think for Fastai, it is important to keep users updated of some kind of status.  In particular if you have things running for hours out of a notebook then seeing status will be important.  I am putting out a PR as a first pass at this.  Time to run is the same, memory use is N * single data copy.  I am curious if you have an example of using a memory map to prevent copying the data to each process? (or at least that is what I think is going on, I am not sure...)

**Comment by kliron** [comment]
I don't really know what you mean by single data copy. My example reads separate files and there is no redundant copies of the same data across the worker processes. If I needed to read from a single csv file, then I would use the `skiprows=...` and `nrows...` arguments of `read_csv` in Pandas to read partitions of rows in each worker, the row ranges being supplied by the main process. Same technique would work with a plaintext file by treating it as a single-column csv and using a character that would never exist in the text as a `sep` value, eg `sep=\b`.

**Comment by bfarzin** [comment]
I saw how yours does that with different files.  I think that something like your indexing is the right way to go, but need to think about how that fits into the rest of the framework for Fastai.

**Comment by parthopdas** [comment]
guys do you know if the above optimizations will resolve [this issue](https://forums.fast.ai/t/getting-cuda-oom-on-lesson3-imdb-notebook-with-a-bs-8/30080/15?u=partho).

So far all the notebooks (with some hyper-param tweakings) have been able to complete on my 1050i+4GBGRAM.

It will be great if the ch4 examples also can be done on my laptop. I think it makes the overall library easier to play with rather than having to get a better gpu or ssh/ts to the cloud.

**Comment by kliron** [comment]
@parthopdas none of the above has anything to do with the GPU or with model training. It has to do with preprocessing of the text, which happens on the CPU and main RAM.

**Comment by parthopdas** [comment]
OK thanks. Do you guys know how i can get the NLP sample to completed on 1050i+4GBGRAM.

With bs=16 & “torch.cuda.empty_cache()” between every save/load I am able to get to unfreezing the last 2 layer. Training with unfreezing the last 3 layers throws GPU OOM.

E.g. reduced batch size? or a simpler architecture? v2018 of the course had a bptt setting - and i see references to that in learner.py - is there a way i can set that while creating the learner?

**Comment by bfarzin** [comment]
Those kinds of questions are best answered on the [fast.ai forums](https://forums.fast.ai/).  Have a look there and post if you still can't find a resolution.

**Comment by sgugger** [comment]
Thanks for all the discussion and great scripts in this thread.  This as all been taken into account and will be solved in the v2 we're currently developping (first dev can be seen [here](https://github.com/fastai/fastai_docs/blob/master/dev_nb/202_tokenizing.ipynb)). v1 will keep the current tokenizer, so there is little point keeping this issue open: it's known and consequently treated.
